Post Featured Image: New design for Replace and Remove buttons

noisysocks opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
Updates the post editor's Featured Image panel to use ~the dropdown-based design that was built in https://github.com/WordPress/gutenberg/pull/42352 but never shipped~ a new design.

## Why?
- The current approach of having _Replace Image_ and _Remove featured image_ buttons below the image looks disoriented.
- ~It gives us room to add an _Upload file_ button.~

## Testing Instructions
1. Create or edit a post.
2. Open the settings sidebar.
3. Add a featured image.

## Screenshots or screencast 

| Before | After |
| - | - |
| <img width="293" alt="Screenshot 2023-05-03 at 11 57 06" src="https://user-images.githubusercontent.com/612155/235820010-c9a35759-90d3-4031-9a2b-f4f41c153153.png"> | <img width="525" alt="Screenshot 2023-05-03 at 11 57 30" src="https://user-images.githubusercontent.com/612155/235820011-4e7cfe77-2524-4bc3-b9e8-b9d111f9137e.png"> |


https://user-images.githubusercontent.com/612155/235820264-f6819b0f-71a8-4859-9d61-33650e6629e8.mp4

noisysocks committed:
Post Featured Image: Use dropdown to add/replace image

noisysocks committed:
Allow popover to shift

noisysocks commented on packages/editor/src/components/post-featured-image/index.js:
```diff
@@ -160,83 +170,111 @@ function PostFeaturedImage( {
+												isOpen,
+												onToggle,
+											} ) => (
+												<Button
```
The amount of indentation here is getting crazy 😅 Will clean it up if we decide to proceed with the design.

noisysocks commented on packages/editor/src/components/post-featured-image/index.js:
```diff
@@ -160,83 +170,111 @@ function PostFeaturedImage( {
+										<Dropdown
+											className="editor-post-featured-image__dropdown"
+											popoverProps={ {
+												placement: 'left-start',
```
This doesn't look the greatest on mobile. Maybe should consider `'overlay'` or `'bottom'` when the viewport is small.

ramonjd replied:
Here's what I'm seeing:

<img width="300" alt="Screenshot 2023-05-03 at 4 24 31 pm" src="https://user-images.githubusercontent.com/6458278/235845017-45aed78a-f60b-476b-8b68-fec13d4a141f.png">

Would it be acceptable to hide the popover in narrower viewports and have a modified version of what's on trunk?

<img width="300" alt="Screenshot 2023-05-03 at 4 29 46 pm" src="https://user-images.githubusercontent.com/6458278/235845762-6b02de7f-1092-41b6-ae7c-52fd7f2b50c0.png">

ramonjd replied:
> Would it be acceptable to hide the popover in narrower viewports and have a modified version of what's on trunk?

Actually, now that I think of it, the replace image button is redundant since the image is a button and triggers the same action.

The remove link could just be a ❌  icon too.

🤷 What do I know.

mtias replied:
For narrower viewports maybe we could do something like this (or with the two buttons below the image, but looking the same)

<img width="250" alt="image" src="https://user-images.githubusercontent.com/548849/236022010-441a45e7-acd7-4a3f-b066-4161728df0a9.png">

aurooba commented:
A thought: what if when you click/select, it transforms into the placeholder with ability to upload, select from the media library, and remove? 

This is not a totally accurate mockup, just a quick one I did in the browswer, but I think it gets the idea acros.
<img width="277" alt="a featured image component resembling the placeholder component for the image block" src="https://user-images.githubusercontent.com/6925260/235837111-277d59f1-17fd-4d65-b481-d3464c5b6fa2.png">

(I'm sure there's a conversation around this elsewhere that I haven't seen so please feel free to dismiss. I'd appreciate a link to context if there is some!)

jasmussen committed:
Fix focus style.

jasmussen committed:
Move the overflow to the button itself.

jasmussen committed:
Apply a max-height.

jasmussen commented:
Nice, thanks for doing that. This is important polish. I feel like it's nearly there. Before:

![before](https://user-images.githubusercontent.com/1204802/235851270-73a8114d-f839-4cdc-94d1-d489e8585c5a.gif)

After:

![featured](https://user-images.githubusercontent.com/1204802/235851320-e9eefe49-7a0d-45d0-99e0-1bbf6b6739bf.gif)

Overall this is much better than before, with a better footprint. A few things, though. In this state, we don't need the flyout. You can drag and drop images, and you can upload inside the media library modal anyway. We can remove the upload and go straight to the media library:

<img width="580" alt="Screenshot 2023-05-03 at 09 09 47" src="https://user-images.githubusercontent.com/1204802/235852112-1bbff57b-a800-4423-b3bc-3969779be08c.png">

In fact, if we remove the upload button, we are left with just two actions, open media library and remove. So can we remove the dropdown entirely, and have two buttons? Just an edit button, and a trash button, as overlay buttons in the top right corner. Clicking the thumbnail just opens the media library:

<img width="599" alt="Screenshot 2023-05-03 at 09 03 53" src="https://user-images.githubusercontent.com/1204802/235852345-1b8c653a-129d-4d21-8f90-a70706cea137.png">

As an overlay button, it's important that it overlays the image, it can't just sit in the white space when the image doesn't fill the area, as it does in square or portrait orientations:

<img width="395" alt="Screenshot 2023-05-03 at 09 06 33" src="https://user-images.githubusercontent.com/1204802/235851574-64d32e5b-d1cb-467e-b89f-548d0737815e.png">

Additionally, portrait featured images get very tall, there's no reason to. I meant to go in and push a fix, so instead of these:

<img width="352" alt="Screenshot 2023-05-03 at 09 16 21" src="https://user-images.githubusercontent.com/1204802/235853179-0a6dd228-fe1d-4270-9f18-ab328bac803d.png">

<img width="313" alt="Screenshot 2023-05-03 at 09 18 16" src="https://user-images.githubusercontent.com/1204802/235854999-02eb585a-50d0-4910-b96f-b2130f4d9b45.png">

we get this:

<img width="351" alt="Screenshot 2023-05-03 at 09 20 34" src="https://user-images.githubusercontent.com/1204802/235855016-8989037d-4586-4ad4-a2ce-be7301a75fef.png">

<img width="343" alt="Screenshot 2023-05-03 at 09 25 41" src="https://user-images.githubusercontent.com/1204802/235855031-67f760d9-c3d3-49b5-ad1f-f2dba7a10bce.png">

I pushed a small fix to the focus style and radius as well.

Note that there is a bottom margin we need to remove:

<img width="957" alt="Screenshot 2023-05-03 at 09 25 49" src="https://user-images.githubusercontent.com/1204802/235855121-40525637-32c4-4e22-93ee-60036eab8f70.png">

It appears here:
<img width="326" alt="Screenshot 2023-05-03 at 09 25 53" src="https://user-images.githubusercontent.com/1204802/235855169-af939c92-6e87-4d45-941f-73e4e939188b.png">

But I can't find it in the codebase. Can you find it and remove it? We might also not need the position relative.

So to summarize:

* This is nearly ready
* Let's add a trash button next to the edit button, and have both the edit button and the thumbnail itself open the media library directly. No need for a dropdown.

Thank you Robert!

ntsekouras commented on packages/editor/src/components/post-featured-image/menu.js:
```diff
@@ -0,0 +1,58 @@
+	onRemoveImage,
+} ) {
+	return (
+		<NavigableMenu className="editor-post-featured-image__menu">
```
Should we use `MediaReplaceFlow` here? The icons are also at the right there and here are on the left.

noisysocks replied:
Not using a dropdown anymore!

aurooba commented:
> can we remove the dropdown entirely, and have two buttons? Just an edit button, and a trash button, as overlay buttons in the top right corner. Clicking the thumbnail just opens the media library.

Popping in as a user to say that I would very much prefer to see ^ this rather than the flyout menu. Some kind of overlay for editing and deleting is a better user experience. Flyout menus that live _outside_ the Inspector visually but are triggered from the Inspector can become problematic very quickly.

mtias commented on packages/editor/src/components/post-featured-image/index.js:
```diff
@@ -160,83 +170,111 @@ function PostFeaturedImage( {
+													removeImageLabel={
+														postType?.labels
+															?.remove_featured_image ||
+														DEFAULT_REMOVE_FEATURE_IMAGE_LABEL
```
Can this be just "Remove" instead of repeating featured image again?

noisysocks committed:
Remove dropdown in favour of a Remove and Edit button

noisysocks committed:
Use same order as before

noisysocks committed:
Tweak labels

noisysocks commented:
Thanks for the feedback @jasmussen. I think I addressed all of it. Let me know if not.

https://user-images.githubusercontent.com/612155/236125417-8126c216-137d-42e4-88df-797386b9598d.mp4

Not sure if you meant for the aspect ratio to be like this on narrow viewports?

<img width="562" alt="Screenshot 2023-05-04 at 16 14 23" src="https://user-images.githubusercontent.com/612155/236125099-3ecfb884-2bc9-47f8-9e38-a64bee7cc8bc.png">

draganescu commented:
I'm glad we got rid of the flyout 🎉

aurooba commented:
If the icons were centred on the bottom, then they would appear on top of the image even if the selected image is a square and doesn't take up the entire featured image div., unless I'm misinterpreting and the plan is to make all images fit the entire featured image div. 🤔

jasmussen commented:
Looks good to me, thanks for the followup!

This looks pretty much ready to go to me, though I think I saw Matías mention Replace/Remove buttons instead of icons? I don't personally have a strong opinion there and can go either way. 

Tabbing works too:

![tabbing](https://user-images.githubusercontent.com/1204802/236204075-4bf987d6-a5df-444a-aa4e-cca03ea54832.gif)

That said, notice at the end of the GIF, there appears to be a focus loss if you invoke either of the two Replace/Remove buttons. I.e. if you tab to the thumbnail itself, press enter, you open the media library modal. Escape closes it again, and sets focus back on the thumbnail. However if you tab to the Edit button, press Enter, then Escape to close the modal, the focus is on `body`, same with the trash button. Can you restore focus back to the thumbnail?

noisysocks committed:
Use glass button design

noisysocks committed:
Fix focus loss

noisysocks committed:
Add fade in / fade out

noisysocks commented:
> This looks pretty much ready to go to me, though I think I saw Matías mention Replace/Remove buttons instead of icons? I don't personally have a strong opinion there and can go either way.

The design he shared in https://github.com/WordPress/gutenberg/pull/50269#discussion_r1184150798? I've gone ahead and implemented it in 258b603. I also added a transition in 4706919 as that felt right. Feel free to tweak the glass effect, transition, or revert the commits altogether if you prefer the icon design.

https://user-images.githubusercontent.com/612155/236355747-3fb8ae48-a375-4b97-8c37-962145dad3cf.mp4

> However if you tab to the Edit button, press Enter, then Escape to close the modal, the focus is on `body`, same with the trash button. Can you restore focus back to the thumbnail?

Fixed in 3909b6e!

jasmussen reviewed (approved):
This tests well for me, and is a substantial improvement over trunk. Let's get it in.

mtias commented:
Let's get it in and iterate, it's nice to see all the QoL improvements to the post settings panel!
